keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,884 @@
+from each cluster in the supercluster.**
+
+
+#### Service Port
+
+A derived service will be accessible with the supercluster IP at the ports
+dictated by child services. If the external properties of service ports for a
+set of exported services don’t match, we won’t know which port is the correct
+choice for a service. For example, if two exported services use different ports
+with the name “http”, which port is correct? What if a service uses the same
+port with different names? As long as there are no conflicts (different ports
+with the same name), the supercluster service will expose the superset of
+service ports declared on its component services. If a user wants to change a
+service port in a conflicting way, we recommend deploying a new service or
+making the change in non-conflicting phases.
+
+
+#### IP Family
+
+Because IPv4 and IPv6 addresses cannot be safely intermingled (e.g. iptables
+rules can not mix IPv4 and IPv6), all component exported services making up a
+supercluster service must use the same `IPFamily`. As with the single cluster
+case - a service’s `IPFamily` is immutable - changing families will require a
+new service to be created.
+
+
+### Component Level Properties
+
+These properties are export-specific and pertain only to the subset of endpoints
+backed by a single instance of each exported service. They may be safely carried
+throughout the supercluster without risk of conflict. We propagate these
+properties forward with no attempt to merge or alter them.
+
+
+#### Session Affinity
+
+Session affinity affects a service as a whole for a given consumer. What would
+it mean for a service to have e.g. client IP session affinity set for half its
+backends? Would sessions only be sticky for those backends, or would there be no
+affinity? If sessions are selectively sticky, we’d expect to see traffic to skew
+toward the sticky subset of endpoints. That said, there’s nothing preventing us
+from applying affinity on a per-slice basis so we will carry it forward.
+
+
+#### TopologyKeys
+
+A `Service`’s `topologyKeys` dictate how endpoints in all `EndpointSlices`
+associated with a given service should be applied to each node. While a single
+`Service` may have multiple `EndpointSlices`, each `EndpointSlice` will only
+ever originate from a single `Service`. `ImportedService` will contain
+label-mapped lists of `topologyKeys` synced from each originating exported
+service. Kube-proxy will filter endpoints in each slice based only on the
+`topologyKeys` defined on the slice’s specific source `Service`.
+
+#### Publish Not-Ready Addresses
+
+Like `topologyKeys` above, we can apply `publishNotReadyAddresses` at the
+per-slice level based on the originating cluster. This will allow incremental
+rollout of changes without any risk of conflict. When true for a cluster, the
+supercluster service DNS implementation must expose not-ready addresses for
+slices from that cluster.
+
+### Test Plan
```

Comment:
You can emulate these scenarios for testing with KIND
cc: @BenTheElder 
I've described it here https://github.com/kubernetes-sigs/kind/pull/1337/files#diff-7906da116e09ff32d7a6c92aaf1113abR63-R69

Reply:
KIND is great :) - I plan to work on this section over the next week

Reply:
Following up here- I've been playing with this for local testing and it works great. Thanks for the tip!

Reply:
FYI that in v0.8.0 (unreleased but soon) nodes will move from the default bridge to a `kind` bridge (to get [docker user defined network features](https://docs.docker.com/network/bridge/#differences-between-user-defined-bridges-and-the-default-bridge), namely the embedded DNS which fixes many problems)

shouldn't affect this much hopefully, all nodes / clusters are still peers on the same docker network. https://github.com/kubernetes-sigs/kind/pull/1337 may need some updates.

Reply:
Thanks, good to know - i'll be following the release